### PR TITLE
Add options for focal loss

### PR DESCRIPTION
**Pull request by guoquan**
Add focal loss to deal with class imbalance

### CODE REVIEW

**Comment by guoquan** [comment]
Sorry, it seems there are some problems. But it works fine in my local copy.
Can anyone send me the error message from CI?

**Comment by matt-gardner** [comment]
You can see it yourself here: http://build.allennlp.org/viewLog.html?buildId=17043&buildTypeId=AllenNLP_AllenNLPPullRequests&tab=buildLog.  Just log in as a guest.

**Comment by guoquan** [comment]
Thanks. Did not see the guest option. Will get back when I figure this out.

**Comment by guoquan** [comment]
Finally here. Not an easy job to make these checkers happy.

File: allennlp/nn/util.py [hunk 1]
```diff
@@ -667,12 +685,51 @@ def sequence_cross_entropy_with_logits(logits: torch.FloatTensor,
...
+                alpha_factor = alpha_factor.view(1)
+                # shape : (2,)
+                alpha_factor = torch.cat([1 - alpha_factor, alpha_factor])
+        else:
+            raise TypeError(('alpha must be float, list of float, or torch.FloatTensor, '
+                             '{} provided.').format(type(alpha)))
+        # shape : (batch, max_len)
+        alpha_factor = torch.gather(alpha_factor, dim=0, index=targets_flat.view(-1)).view(*targets.size())
+        weights = weights * alpha_factor
+    else:
```
Comment (matt-gardner) [inline allennlp/nn/util.py]: I don't think you need this `else`.

File: allennlp/nn/util.py [hunk 2]
```diff
@@ -667,12 +685,51 @@ def sequence_cross_entropy_with_logits(logits: torch.FloatTensor,
...
+    # focal loss coefficient
+    if gamma:
+        # shape : (batch * sequence_length, num_classes)
+        probs_flat = log_probs_flat.exp()
+        # shape : (batch * sequence_length,)
+        probs_flat = torch.gather(probs_flat, dim=1, index=targets_flat)
+        # shape : (batch * sequence_length,)
+        focal_factor = (1. - probs_flat) ** gamma
+        focal_factor = focal_factor.view(-1, 1)
+    else:
```
Comment (matt-gardner) [inline allennlp/nn/util.py]: I think you can similarly modify `weights` as the last line of the `if` above, and not need this `else`, or need to reference `focal_factor` down below.

File: allennlp/tests/nn/util_test.py [hunk 3]
```diff
@@ -661,6 +662,7 @@ def test_sequence_cross_entropy_with_logits_averages_batch_correctly(self):
         # Batch has one completely padded row, so divide by 4.
         assert loss.data.numpy() == vector_loss.sum().item() / 4
 
+    @flaky # some time "not almost equal to 7 decimals"
```
Comment (matt-gardner) [inline allennlp/tests/nn/util_test.py]: No need for the comment here; that's already the definition of `flaky`.  I'm not sure why the change you made would make this test flaky, though, while the ones below aren't.

File: allennlp/tests/nn/util_test.py [hunk 4]
```diff
@@ -681,6 +683,106 @@ def test_sequence_cross_entropy_with_logits_averages_token_correctly(self):
...
+
+        loss = util.sequence_cross_entropy_with_logits(tensor, targets, weights, gamma=gamma)
+
+        correct_loss = 0.
+        for logit, label in zip(tensor.squeeze(0), targets.squeeze(0)):
+            p = torch.nn.functional.softmax(logit, dim=-1)
+            pt = p[label]
+            correct_loss += pt.log() * (1 - pt) ** gamma
+        # Average over sequence.
+        correct_loss = - correct_loss / length
```
Comment (matt-gardner) [inline allennlp/tests/nn/util_test.py]: Slightly clearer to subtract inside the loop and just average here.  Same for all tests below.

File: allennlp/tests/nn/util_test.py (hunk 3 above)
Comment (guoquan) [reply allennlp/tests/nn/util_test.py]: It just happened to fail at a very early stage I try to modify the loss. Did not notice if it is related to my changes. It happens once out of 10 rounds. It really looks irrelevant. I just add this to stop the unit test complaining.
My test is using token-average, as in smooth label test. So it is not affected by the unstable problem when using batch-average, in the above flaky test.

However, I also see once test with the token-average complaint "not almost equal to 7 decimals", out of maybe 50+ times local and TC tests. Please refer to [# 7006](http://build.allennlp.org/viewLog.html?buildId=17137) in TC. Need `@flaky` for it also? This may imply all other tests rely on token-average could be flaky. There are 8-9 calls to token-average which is the default. So I will estimate the chance of flaky is ~1/500.

File: allennlp/tests/nn/util_test.py (hunk 4 above)
Comment (guoquan) [reply allennlp/tests/nn/util_test.py]: Great. Thanks.

File: allennlp/nn/util.py (hunk 1 above)
Comment (guoquan) [reply allennlp/nn/util.py]: Yes. I used to have something like `focal_factor` but combining into `weights` looks better. Removed this `else`.

File: allennlp/nn/util.py (hunk 2 above)
Comment (guoquan) [reply allennlp/nn/util.py]: Great. Update.

File: allennlp/tests/nn/util_test.py (hunk 3 above)
Comment (guoquan) [reply allennlp/tests/nn/util_test.py]: Sarcasm is that I get flaky immediately with this build http://build.allennlp.org/viewLog.html?buildId=17245. Alright, I will add `@flaky` to all of those.

**Comment by guoquan** [comment]
It seems moving where to combine the weight brings in new problem. Will look at it later.

File: allennlp/nn/util.py (hunk 2 above)
Comment (guoquan) [reply allennlp/nn/util.py]: Hi @matt-gardner. I realize there is a problem here because when calculating the average, "token", "batch", or even "none", it relies on `weights.sum()`. Do you think `gamma` (and also `alpha`) should go into the denominator of average `weights.sum()`? It is a kind of design decision, rather than correctness, of how to calculate the average here. What do you think?

File: allennlp/tests/nn/util_test.py (hunk 3 above)
Comment (guoquan) [reply allennlp/tests/nn/util_test.py]: It seems no problem this part as I discuss in the above conversation. The question here is if you want a minimum `@flaky` flags? If so I remove not necessary ones.

Comment (matt-gardner) [reply allennlp/tests/nn/util_test.py]: I don't see a problem with having a `@flaky` decorator, I was just surprised it's necessary.

File: allennlp/nn/util.py (hunk 2 above)
Comment (matt-gardner) [reply allennlp/nn/util.py]: `alpha` and `gamma` have the effect of changing the weighting on each of the instances / tags in the batch, so it will affect the direction of the gradients.  Changing the denominator in `weights.sum()` changes all directions of the gradient equally, so including `alpha` and `gamma` in the sum effectively just lowers the learning rate.  Right?  If I'm understanding this correctly, I'd probably compute the sum before multiplying by the `alpha` and `gamma` factors, so the effective learning rate doesn't change as much.

Comment (guoquan) [reply allennlp/nn/util.py]: Right. The outcome would be changing the learning rage if they are included. If this is the case, I can calculate the `sum()` before involving them.
